forte/processors/base/data_augment_processor.py

```diff
@@ -16,20 +16,196 @@
+    are the outputs. Each of the span has start and end index.
+    The old_spans and new_spans are anchors for the mapping,
+    because we depend on them to determine the position change of the
+    index.
+
+    Given an index, the function will find its the nearest
+    among the old spans before the index, and calculate the difference
+    between the position of the old span and its corresponding new span.
+    The position change is then applied to the input index. An updated
+    index is then calculated and returned.
+
+    An inserted span might be included as a part of another span.
+    For example, given a sentence "I love NLP.", if we insert a
+    token "Yeah" at the beginning of the sentence(index=0), the Sentence
+    should include the new Token, i.e., the Sentence will have a
+    start index equals to 0. In this case, the parameter is_inclusive
+    should be True. However, for another Token "I", it should not include
+    the new token, so its start index will be larger than 0.
+    The parameter in_inclusive should be False.
+
+    The input index could be the start or end index of a span, i.e., the
+    left or right boundary of the span. If there is an insertion in the span,
+    we should treat the two boundaries in different ways. For example,
+    we have a paragraph with two sentences "I love NLP! You love NLP too."
+    If we append another "!" to the end of the first sentence, when modifying
+    the end index of the first Sentence, it should be pushed right to include
+    the extra exclamation. In this case, the is_begin is False. However, if
+    we prepend an "And" to the second sentence, when modifying the start index
+    of the second Sentence, it should be pushed left to include the new Token.
+    In this case, the is_begin is True.
+
+    Args:
+        index (int): The index to map.
+        old_spans (SortedList): The spans before replacement. It should be
+            a sorted list in ascending order.
+        new_spans (SortedList): The spans after replacement. It should be
+            a sorted list in ascending order.
+        is_begin (bool): True if the input index is the start index of a span.
+        is_inclusive (bool): True if the span constructed by the aligned
+            index should include inserted spans.
+    Returns:
+        The aligned index.
+
+    If the old spans are [0, 1], [2, 3], [4, 6],
+    the new spans are [0, 4], [5, 7], [8, 11],
+    the input index is 3, and there are no insertions,
+    the algorithm will first locate the last span with
+    a begin index less or equal than the target index,
+    ([2,3]), and find the corresponding span in new spans([5,7]).
+    Then we calculate the delta index(7-3=4) and update our
+    input index(3+4=7). The output then is 7.
+
+    Note that when the input index locates inside the old spans,
+    instead of on the boundary of the spans, we just keep the offset
+    of the input index to the old span it was in unchanged. In the
```

Comment:
we compute the return index so that it maintains the same offset to the begin of the span it belongs to.